Introduce st.navigation and st.Page

kmcgrady opened the pull request:
## Describe your changes

This PR is the culmination of all of the work associated with Multipage apps V2. Specifically this PR includes:

- abstractions on the frontend and backend for navigation and page management respectively.
- command for st.navigation and st.Page

## GitHub issues
closes #8759

## Testing Plan

- JS and Python End to End tests are included
- E2E Tests are included for the MPA v2 features

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

kmcgrady committed:
Migrate frontend navigation to App Navigation Abstraction (#8631)

## Describe your changes

Combine all logic of the navigation (page url updates, mpa management),
and put it in a class (V1AppNavigation). We create a layer above it that
essentially pulls from the V1AppNavigation as an overengineering effort
to set up for AppNavigation V2

## Testing Plan

- Original tests should pass
- unit tests for AppNavigation.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Refactor page management to a PagesManager class (#8639)

## Describe your changes

Migrates pages are primarily managed in the `source_util` module to a
`PagesManager` class. Because V2 can have a dynamic set of pages per
session, the Pages Manager instance will live on the `AppSession`. For
V1, we will leverage static variables/methods for page watching.

## Testing Plan

- Original Unit tests are applied (with patches adjusted)
- Unit tests are applied for the PagesManager component

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Address Pages Manager lock behavior and types (#8650)

## Describe your changes

From #8639, @raethlein made a good claim about the lock as well as some
types to make things clear, so I addressed those here.

## Testing Plan

- Type checks should pass
- Existing tests should pass with lock behavior (we don't usually manage
concurrency testing at the moment)

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Introduce concept of active script hash (#8656)

## Describe your changes

MPA v2 will introduce the concept of an active script hash where the
active script is the current script running. This change does the
following:

- Adds the concept of Active Script Hash to the Pages Manager. By
default, it returns the current page hash and doesn't set anything.
- All widgets will leverage the Active Script Hash in calculating their
id. We do this so that widgets can be associated with the script (and
avoid duplicate issues where possible).
- We will provide the active script hash as part of the forward message
metadata. That way, we can associate the active script hash with every
call (this is especially important for `Delta` messages.
- Frontend will store the active script hash in the nodes

## Testing Plan

- Added a python unit test for the active script hash setting in the
forward message.
- Open to ideas on how to test the widget id updates, but I think it's
not quite feasible.
- Frontend tests storage of active script hash in nodes

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Include Main Script Hash info in App Nodes (#8657)

## Describe your changes

MPA v2 requires to understand which code is associated with the main
script (or common script) or not. We can send the main script hash in
the NewSession

## Testing Plan

- Updated unit tests to include the main script hash
- Added a test that verifies the main script hash is set correctly

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Simplify updateReport call to remove properties (#8667)

## Describe your changes

Confirmed that the properties associated with the `updateReport` call
are not needed.

## Testing Plan

- Added a unit test to ensure the value is sent on new session

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Introduce a strategy design pattern for PagesManager and AppNavigation (#8658)

## Describe your changes

MPA v2 will present different implementations from previous versions. We
will adopt a strategy design pattern in order to separate implementation
from MPA v1 or v2.

## Testing Plan

- Original unit tests from this feature branch should still apply

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Add st.navigation, st.Page and associated V2 strategies (#8673)

## Describe your changes

This change includes `st.Page` and `st.navigation` commands bringing us
closer to the finish point for MPA v2.

## Testing Plan

- Added associated unit tests
- E2E Tests will be needed to test MPA end-to-end, but they will not be
included in this PR to the feature branch.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Convert Navigation position string type to enum (#8693)

## Describe your changes

There was a good recommendation to move position to an enum. Let's move
it before committing the feature.

## Testing Plan

- Updated Unit Tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Make the main script hash optional in AppRoot.empty() (#8694)

## Describe your changes

Others (Notebooks) uses `AppRoot.empty()`. To make things easier, I make
the `mainScriptHash` keyword only.

## Testing Plan

- Existing tests should be fine enough

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Add support for st.Page for switch_page and page_link (#8692)

## Describe your changes

Title says it all. Ensure `st.Page` is an argument to `switch_page` and
`page_link`

## Testing Plan

- Unit Tests for `page_link`
- E2E Tests will be added for `switch_page`

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
set_page_config should override mpa-v2 Page properties (#8691)

## Describe your changes

If `st.set_page_config` is called, we should disregard the page
information in the navigation call.

## Testing Plan

- JS Unit Tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Retrieve pages for new session when ready to send the event (#8702)

## Describe your changes

We have found flaky situations where pages are set via `st.navigation`
before the NewSession call is sent. This causes the NewSession message
to retrieve pages it should not be expecting from the PagesManager. The
easy solution here is to send the pages on the Script Started Event so
that the NewSession has the pages at the time of the event.

I believe there's a larger discussion on how we should be managing this.
Likely, we should ensure information is happening at the right moment.

## Testing Plan

- Python unit test updated. This is difficult to reproduce and therefore
difficult to test.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Don't notify pages change for MPAv2 (#8698)

## Describe your changes

Page Changes in MPAv2 are done via the `st.navigation` call. Informing
of Pages Change is not helpful for the UI display. We also came to the
conclusion that the concept of cache pages did not matter for
PagesManager as well. So I moved everything back to source_util to
produce a cleaner system.

In testing this, I noticed some inconsistencies due to the fact that the
default strategy of the PagesManager is not resetting. So I did the
following:

1. I moved the DefaultStrategy reset to be set on every every test to
make things easy.
2. I then found out that when MPA switches from V1->V2 the initial
script information (page hash and page name) got lost, which suggested
this should be in the root PagesManager and not tied to just the V2
strategy.
3. I found a better name than "initial" I would describe it better as
the "intent", which is to indicate to the PagesManager a page is being
requested.

## Testing Plan

- Python Tests for both strategies

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Support Navigation Sections in the Sidebar for MPAv2 (#8681)

## Describe your changes

These changes allow SidebarNav to display NavSections.

<img width="380" alt="image"
src="https://github.com/streamlit/streamlit/assets/69432/39f7a879-931b-49e8-8e47-f9a4dc1015fe">

- The key difference in the screenshots is the absence of the x
- There's some very minor margin changes


## Testing Plan

- JS unit tests are provided
- E2E Tests are updated to reflect the new images for MPA v1 mainly.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

raethlein commented on e2e_playwright/__snapshots__/linux/hello_app_test/hello_app-animation_demo_page[chromium].png:
It looks like the side-bar items moved closer to the top. Just sanity-checking whether this visual change is intended?

kmcgrady replied:
Double checked with Jessi, and the answer is yes!

raethlein commented on frontend/app/src/App.tsx:
```diff
@@ -151,9 +152,10 @@ interface State {
   appPages: IAppPage[]
+  navSections: string[]
   currentPageScriptHash: string
+  mainScriptHash: string
```
it might make sense to add a comment here explaining the difference real quick between `mainScriptHash` and `currentPageScriptHash` or a link to the explanation

raethlein commented on frontend/app/src/App.tsx:
```diff
@@ -653,10 +667,23 @@ export class App extends PureComponent<Props, State> {
+  handleLogo = (logo: Logo, metadata: ForwardMsgMetadata): void => {
+    this.setState(
+      {
+        elements: this.pendingElementsBuffer.appRootWithLogo(logo, metadata),
```
I am still a little bit surprised that we have to create a new AppRoot with elements etc. in order to set the logo 🤔 High likelihood I don't understand it correctly, but if you feel the same we should probably create a tech debt item and refactor this (there is anyways a refactor opportunity here to make the class-based component a functional one)

kmcgrady replied:
Yea, AppRoot is immutable. Every delta generates a new AppRoot. I think there's opportunity to make it cleaner, but I kinda like the immutability factor here.

raethlein replied:
Oh I wasn't aware that every delta message creates a new AppRoot. I have to spend a little bit more time looking into this 🙂

raethlein commented on frontend/app/src/App.tsx:
```diff
@@ -1038,17 +1027,14 @@ export class App extends PureComponent<Props, State> {
     const hasAnchor = document.location.toString().includes("#")
-    const isSamePage =
-      targetAppPage?.pageScriptHash === this.state.currentPageScriptHash
+    const isSamePage = targetAppPage.pageScriptHash === currentPageScriptHash
```
in the old version it looks like  `targetAppPage?` could have been `null` and the [next line ](https://github.com/streamlit/streamlit/pull/8744/files#diff-1b95d4e0dda1e5928f99c8303599bd811f140fdd40e028fcfca3077f802d6210R1039) still indicates that this might be possible. So should this be `targetAppPage?.pageScriptHash`?

kmcgrady replied:
Yes, I may have removed it cause I forced the types to understand how MPA works. It might be safe to re-include this just for safety

raethlein commented on frontend/app/src/App.tsx:
```diff
@@ -1193,11 +1179,20 @@ export class App extends PureComponent<Props, State> {
+        const activeWidgetIds = new Set(
+          Array.from(this.state.elements.getElements())
+            .map(element => getElementWidgetID(element))
+            .filter(notUndefined)
```
nit: could you add a comment here please about in which scenarios elements can be undefined despite being in the elements list? Or is it just a precautious check?

kmcgrady replied:
Yes. Will do. For context, this is all elements, and not all elements are widgets. So this is trying to get all the widget ids. For non-widgets, it will just return null/undefined.

raethlein replied:
ahh ok, thanks! yeah I think adding that line as a comment here would be great

raethlein commented on lib/streamlit/commands/navigation.py:
```diff
@@ -0,0 +1,189 @@
+    default_page = None
+    pagehash_to_pageinfo: dict[PageHash, PageInfo] = {}
+
+    # This nested loop keeps track of three things:
+    # 1. the default page
+    # 2. the pagehash to pageinfo mapping
```
nit: "keeps track of three things" but then you list only 2 😁 what's the third?

kmcgrady replied:
Ah. I removed the third at the last minute (I should stop counting things 😆 )

raethlein commented on lib/streamlit/commands/navigation.py:
```diff
@@ -0,0 +1,189 @@
+            if script_hash in pagehash_to_pageinfo:
+                raise StreamlitAPIException(
+                    f"Multiple Pages specified with title {page.title}. "
+                    "Page titles should be unique. If not specified, "
```
nit: "Page titles should be unique" -> "Page titles must be unique"

kmcgrady replied:
I think this is a weird error. The error case is: "when there's duplicate script hashes these two pages share all similar information" The way to differentiate would be the title. So not every page title HAS to be unique, but if there's no other way to define the script hash, the titles need to be changed.

In fairness, this should be a relatively rare exception.

raethlein commented on lib/streamlit/commands/navigation.py:
```diff
@@ -0,0 +1,189 @@
+                raise StreamlitAPIException(
+                    f"Multiple Pages specified with title {page.title}. "
+                    "Page titles should be unique. If not specified, "
+                    "title is inferred from the file path or callable name."
```
nit: "title is" -> "the title is"

raethlein commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -0,0 +1,341 @@
+from pathlib import Path
+from typing import Any, Callable, Final, Optional, Type
+
+import streamlit.source_util as source_util
```
it looks like you should add the required function from source_util here: https://github.com/streamlit/streamlit/pull/8744/files#diff-58c845510f4d060aa8b239f4dce3cab3d212a5324bac9e0131a3afa53b509efbR26

kmcgrady replied:
Hmm. I would probably disregard for this example. There are numerous functions needed to be imported (like `get_pages`), and they can create confusion. I can rename the imports, but they get separated into different imports. This just seems like the cleanest IMO.

lukasmasuch replied:
Why not import like: `from streamlit import source_util`? Also, I think the warning could be cleaned up by moving the types below into `TYPE_CHECKING`, e.g.:

```python
if TYPE_CHECKING:
    from streamlit.source_util import PageHash, PageInfo, PageName, ScriptPath
```

raethlein commented on proto/streamlit/proto/AppPage.proto:
```diff
@@ -29,4 +29,9 @@ message AppPage {
   string page_name = 2;
   string icon = 3;
+
+  // A feature for MPA v2 and inform the frontend what's the default page
```
nit: I think it should be "A feature for MPA v2 and inform" -> "A feature for MPA v2 to inform"

raethlein commented on lib/tests/streamlit/web/server/routes_test.py:
```diff
@@ -166,13 +167,22 @@ def test_parse_url_path_200(self):
             assert r.code == 200
 
-    def test_parse_url_path_404(self):
+    def test_most_404_return_default(self):
```
what do you mean with "most" here? A better name might be `test_nonexistent_urls_return_default_page`, what do you think?

raethlein commented on lib/tests/streamlit/runtime/pages_manager_test.py:
```diff
@@ -0,0 +1,202 @@
+# pytest fixture is so useful for writing this test it's worth having the
+# slight inconsistency.
+@patch("streamlit.source_util._cached_pages", new=None)
+def test_get_initial_active_script_v1(tmpdir):
```
cool fixture!
nit: `tmpdir` seems to be a [py.path.local](https://py.readthedocs.io/en/latest/path.html#py-path-local-local-file-system-path) object according to the [pytest fixture docs](https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmpdir-fixture), in case you want to type it

kmcgrady replied:
Hmmmm...I know this is silly, but py is not in our dependencies. I'm sure it would work if I import it (cause pytest must require it), but I don't want to add test dependency and have that break our tests. Open to other ideas if you feel this is important.

raethlein replied:
that makes 100% sense to me to not use any transient dependencies!

raethlein commented on lib/tests/streamlit/navigation/page_test.py:
```diff
@@ -0,0 +1,83 @@
+from tests.delta_generator_test_case import DeltaGeneratorTestCase
+
+
+class PagesManagerTest(DeltaGeneratorTestCase):
```
[This test class](https://github.com/streamlit/streamlit/pull/8744/files#diff-197713706126976de58166557ddedba2d248b1865fc71f13e413b0956931de90R26) is also called `PagesManagerTest`, which could be confusing in the output. It might be worth to rename this one here

raethlein commented on lib/tests/streamlit/navigation/page_test.py:
```diff
@@ -0,0 +1,83 @@
+            st.Page("page.py", icon="hello world")
+
+    def test_valid_icon(self):
+        """Test that passing an valid icon does not raise an exception."""
```
nit: "passing an valid" -> "passing a valid"

raethlein commented on lib/tests/streamlit/navigation/page_test.py:
```diff
@@ -0,0 +1,83 @@
+            != st.Page(lambda: True, title="Title 2")._script_hash
+        )
+
+    def test_page_run_cannot_be_run_automatically(self):
```
nit: "automatically" might not be the best word here, as nothing happens automatically here. What do you think about "cannot_be_run_standalone" or similar and extend the docstring with the comment I have seen somewhere that only the page that is returned by the navigation can be run

raethlein commented on lib/tests/streamlit/navigation/page_test.py:
```diff
@@ -0,0 +1,83 @@
+        with pytest.raises(StreamlitAPIException):
+            st.Page("page.py").run()
+
+    def test_page_run_cannot_be_run_if_ordained(self):
```
"test_page_run_cannot_be_run_if_ordained" -> "test_page_run_can_be_run_if_ordained"

raethlein commented on lib/streamlit/runtime/scriptrunner/script_run_context.py:
```diff
@@ -60,9 +61,9 @@ class ScriptRunContext:
     uploaded_file_mgr: UploadedFileManager
     main_script_path: str
-    page_script_hash: str
     user_info: UserInfo
     fragment_storage: "FragmentStorage"
+    pages_manager: "PagesManager"
```
curious: when do we use the TYPE_CHECKING only typing versus the other? To me it seems to be a little bit arbitrary

kmcgrady replied:
My interpretation is that the TYPE_CHECKING is used for imports and to avoid circular dependencies.

raethlein commented:
I am excited to see MPAv2 landing! 
I think for future reference it would be great if you could link all PRs that went into this PR here in the description; this can help to find conversations etc. if we need to revisit anything again.
Also, since this is the official PR going into develop, it might make sense to add a little bit more details about the feature and new commands etc. to the description.

kmcgrady committed:
Add E2E tests for MPA V2 (#8743)

lukasmasuch commented on e2e_playwright/__snapshots__/linux/mpa_basics_test/mpa-sidebar_nav[dark_theme-webkit].png:
Just to double-check, the color change of the selected page is intentional?

kmcgrady replied:
Double checked with Jessi, and the answer is yes!

lukasmasuch commented on e2e_playwright/multipage_apps_v2/mpa_v2_basics.py:
```diff
@@ -0,0 +1,95 @@
+
+
+def page_8():
+    st.header("Page 8")
```
Curious question: If I import a module that contains the app code within this function, it would probably not work reliably or (since the module import is only triggered once)? E.g.:

```python
def page_8():
    from my_app import my_nice_page
```

 I would need to executed the code via __main__ or some other function?

kmcgrady replied:
Just to make sure I understand. Are you saying the st calls are in `my_app` or in `my_nice_page` as a function? I imagine the `my_nice_page` as a function is not a problem because the function will always be executed. If you are saying in `my_app`. I can test it, but I would imagine it would follow the same trend of the main script. I can try both, but I want to understand the concern.

lukasmasuch replied:
No concern, just a question... Nothing to do here. I will just try it out after the release

lukasmasuch commented on e2e_playwright/multipage_apps_v2/mpa_v2_basics_test.py:
```diff
@@ -0,0 +1,336 @@
+    # Expand the nav
+    view_button.click(force=True)
+    # We apply a quick timeout here so that the UI has some time to
+    # adjust for the screenshot after the click
```
nit: it doesn't seem the comment is still relevant since there isn't a timeout, or?

sfc-gh-jcarroll commented on lib/streamlit/navigation/page.py:
```diff
@@ -0,0 +1,168 @@
+        if self.icon:
+            validate_icon_or_emoji(self.icon)
+
+        self.default: bool = default
```
This ends up being exposed in st.help() and Debbie will probably want to document it. Would it make sense to be `self._default` instead?

I'm not 100% opposed to exposing it but it wasn't in the spec, just thinking through it out loud. Thoughts?

kmcgrady replied:
Happy to go conservative and make it internal to start and we can think about it more.

sfc-gh-jcarroll commented on lib/streamlit/navigation/page.py:
```diff
@@ -0,0 +1,168 @@
+                )
+
+        self._page: Path | Callable[[], None] = page
+        self.title: str = title or inferred_name
+        self.icon: str = icon or inferred_icon
```
Should these (and `page_url`) be properties to internal variables, to prevent people from manipulating them after initialization and potentially causing weird behavior?

kmcgrady replied:
Certainly possible to prevent the idea of changing. I don't see a problem with users changing it, but we can be conservative here as well.

vdonato reviewed (approved):
Mostly giving this a rubber stamp after skimming the code to sanity check it since most of the effort to review this code was already done in PRs into the feature branch, but this LGTM

lukasmasuch commented on frontend/cypress/snapshots/linux/2x/hostframe.spec.js/theme-message-before.snap.png:
nit: is this expected that these snapshots don't have a closing icon in the sidebar?

kmcgrady replied:
Yes. The close button will appear on hover.

lukasmasuch replied:
Oh, i see 👍

lukasmasuch commented on lib/streamlit/elements/widgets/button.py:
```diff
@@ -681,38 +682,45 @@ def _page_link(
+                page_link_proto.label = page.title
+        else:
+            # Handle external links:
+            if page.startswith("http://") or page.startswith("https://"):
```
nit: we have a slightly more advanced method to check for URLs in url_utils that could be used here:

https://github.com/streamlit/streamlit/blob/00ac86c01eae06ff00fa57207ebce3612ddb5382/lib/streamlit/url_util.py#L71C12-L72C1

lukasmasuch commented on lib/streamlit/elements/widgets/button.py:
```diff
@@ -681,38 +682,45 @@ def _page_link(
+            if page.startswith("http://") or page.startswith("https://"):
+                if label is None or label == "":
+                    raise StreamlitAPIException(
+                        f"The label param is required for external links used with st.page_link - please provide a label."
```
super nit: this doesn't have to a functional string

lukasmasuch commented on lib/streamlit/web/server/routes.py:
```diff
@@ -41,10 +41,15 @@ def allow_cross_origin_requests() -> bool:
+    def initialize(
+        self,
+        path: str,
+        default_filename: Optional[str] = None,
```
nit: `str | None`

lukasmasuch replied:
nit: `str | None`

lukasmasuch commented on lib/streamlit/navigation/page.py:
```diff
@@ -0,0 +1,168 @@
+        inferred_icon = ""
+        if isinstance(page, Path):
+            inferred_icon, inferred_name = page_icon_and_name(page)
+        elif title is None:
```
nit: we could also check if its a callable here and if not raise an exception that the type is not supported.

lukasmasuch commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -0,0 +1,341 @@
+
+    def get_initial_active_script(
+        self, page_script_hash: PageHash, page_name: PageName
+    ) -> Optional[PageInfo]:
```
nit:  `PageInfo | None`

lukasmasuch commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -0,0 +1,341 @@
+            or self.pages_manager.main_script_hash,  # Default Hash
+        }
+
+    def get_page_script(self, fallback_page_hash: PageHash) -> Optional[PageInfo]:
```
nit: `PageInfo | None`

lukasmasuch commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -0,0 +1,341 @@
+        self,
+        callback: Callable[[str], None],
+    ) -> Callable[[], None]:
+        # V2 strategy does not handle any pages changed event
```
nit: does this mean that for V2 we would not detect changes to page scripts? Does this have any impact on local development?

kmcgrady replied:
It does not. It just does not call the pages_changed since the navigation command will always inform of the new pages.

lukasmasuch replied:
So, it's just about new pages and not about changes within the page scripts?

kmcgrady replied:
Correct!

lukasmasuch commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -0,0 +1,341 @@
+
+        self.pages_strategy.set_pages(pages)
+
+    def get_page_script(self, fallback_page_hash: PageHash = "") -> Optional[PageInfo]:
```
nit: `PageInfo | None`

lukasmasuch commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -0,0 +1,341 @@
+
+    DefaultStrategy: Type[PagesStrategyV1 | PagesStrategyV2] = PagesStrategyV1
+
+    def __init__(self, main_script_path, script_cache=None, **kwargs):
```
nit: maybe annotate the types in here?

lukasmasuch commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -0,0 +1,341 @@
+        )
+
+    @contextlib.contextmanager
+    def run_with_active_hash(self, page_hash):
```
nit: maybe annotate `page_hash`

lukasmasuch commented on lib/streamlit/navigation/page.py:
```diff
@@ -0,0 +1,168 @@
+        if isinstance(self._page, Path):
+            h = calc_md5(str(self._page))
+        else:
+            h = calc_md5(self.title)
```
What happens if two pages have the same title? Is that possible?

kmcgrady replied:
If the page is also a callable it will be a duplicate and will throw an exception. I think adding `url_path` will simplify this because it's an easy unique constraint.

lukasmasuch commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -0,0 +1,341 @@
+    def main_script_hash(self) -> PageHash:
+        return self._main_script_hash
+
+    @property
+    def intended_page_name(self) -> PageName | None:
+        return self._intended_page_name
+
+    @property
+    def intended_page_script_hash(self) -> PageHash | None:
+        return self._intended_page_script_hash
+
+    def get_main_page(self) -> PageInfo:
+        return {
+            "script_path": self._main_script_path,
+            "page_script_hash": self._main_script_hash,
+        }
+
+    def get_current_page_script_hash(self) -> PageHash:
+        return self._current_page_hash
+
+    def set_current_page_script_hash(self, page_hash: PageHash) -> None:
+        self._current_page_hash = page_hash
+
+    def get_active_script_hash(self) -> PageHash:
+        return self.pages_strategy.get_active_script_hash()
+
+    def set_active_script_hash(self, page_hash: PageHash):
+        return self.pages_strategy.set_active_script_hash(page_hash)
+
+    def set_script_intent(
+        self, page_script_hash: PageHash, page_name: PageName
+    ) -> None:
+        self._intended_page_script_hash = page_script_hash
+        self._intended_page_name = page_name
+
+    def get_initial_active_script(
+        self, page_script_hash: PageHash, page_name: PageName
+    ) -> Optional[PageInfo]:
+        return self.pages_strategy.get_initial_active_script(
+            page_script_hash, page_name
```
nit: maybe it a quick docstring to those methods. e.g. just from looking at it its not super obvious what the difference of `get_active_script_hash` and `get_current_page_script_hash` is.

lukasmasuch commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -0,0 +1,341 @@
+
+    def get_initial_active_script(
+        self, page_script_hash: PageHash, page_name: PageName
+    ) -> Optional[PageInfo]:
```
nit: `PageInfo | None`

lukasmasuch commented on lib/streamlit/commands/execution_control.py:
```diff
@@ -96,7 +96,7 @@ def experimental_rerun() -> NoReturn:
 
 @gather_metrics("switch_page")
-def switch_page(page: str) -> NoReturn:  # type: ignore[misc]
+def switch_page(page: str | StreamlitPage) -> NoReturn:  # type: ignore[misc]
```
nit: maybe also support `Path` here for consistency?

lukasmasuch commented on lib/streamlit/commands/navigation.py:
```diff
@@ -0,0 +1,189 @@
+def navigation(
+    pages: list[StreamlitPage] | dict[SectionHeader, list[StreamlitPage]],
+    *,
+    position: Literal["sidebar"] | Literal["hidden"] = "sidebar",
```
nit: `Literal["sidebar", "hidden"]`

lukasmasuch commented on lib/streamlit/commands/navigation.py:
```diff
@@ -0,0 +1,189 @@
+        A list of `st.Page` objects or a dictionary where the keys are section
+        headers and the values are lists of `st.Page` objects.
+
+    position: str
```
nit: "sidebar" or "hidden"

lukasmasuch commented on lib/streamlit/commands/navigation.py:
```diff
@@ -0,0 +1,189 @@
+                "page_name": page.title,
+                "icon": page.icon,
+                "script_path": script_path,
+                "url_pathname": page.title.replace(" ", "_"),
```
nit: should we also lowercase the path name? Case-sensitive path names look a bit ugly. Also, it might make sense to put the logic into a shared utils method - e.g., `title_to_url_path` - since it's at least done one more time below. And its easy to otherwise lose track of this and only apply the change at one point.

kmcgrady replied:
I think this is a good question, but I think it mimics a similar behavior to MPAv1 and encourage people to use a follow-up `url_path` keyword param to modify it to their liking.

lukasmasuch commented on lib/streamlit/web/server/routes.py:
```diff
@@ -59,28 +64,21 @@ def set_extra_headers(self, path: str) -> None:
-            url_path = "/".join(url_parts[1:])
-
-        return super().parse_url_path(url_path)
+    def validate_absolute_path(self, root: str, absolute_path: str) -> Optional[str]:
```
nit: `str | None`

lukasmasuch commented on lib/streamlit/web/server/routes.py:
The changes in this file seem to be a bit independent. Maybe it makes sense to extract this into a dedicated PR and get it merged into `develop`

lukasmasuch reviewed (commented):
LGTM 👍 I added a couple of nits, but I quickly skipped through a couple of parts of the PR.

It might be good to deactivate the `hello_app_test` temporarily until someone has rewritten the hello app to use MPA v2. Always having them rerun once doesn't feel like a great state and slows down our CI job.  Probably just renaming it to `hello_app_test_deactivated.py` would ignore it from running. 

As a follow-up after the MPA v2 release, I still think it would be doable to remove V1 and "simulate" the old behavior with the v2 implementation.

kmcgrady committed:
Address feedback on MPAv2 branch (#8777)

## Describe your changes

Feedback has beens received on #8744, and instead of issuing a commit, I
created a PR with the changes associated. Changes include:

- Additional comments for clarity
- Renaming of tests/comments for clarity
- Making fields internal for `st.Page` and exposing as properties for
read-only access.
- Not sending the full title but the page name in host communication
(avoids double "Streamlit" being added
- Fixing an (v1 -> v2) edge case that requires the system to know MPA
v2.
- Ensure pages are watched after every successful run.

## Testing Plan

- Unit tests/E2E tests are updated and should pass.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Merge branch 'develop' into feature/mpa-v2

kmcgrady committed:
Address type issues

kmcgrady committed:
Specify url path in st.Page (#8778)

## Describe your changes

Adds `url_path` as a feature of `st.Page`. This change does three
things:
1. Offers the ability to specify the `url_path` (it will strip any
preceding slashes).
2. If not specified, the `url_path` will be inferred first from the file
or function name, then from the title
3. The script hash is based on the url_path, which simplifies
everything. Note: I intentionally kept the default page as defined
rather than a hash of "" to avoid any potential issues.

## Testing Plan

- Unit Tests and E2E Tests are updated and include tests for the
url_path.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady committed:
Make sure fragments maintain active page script hash on subsequent reruns (#8802)

## Describe your changes

Fragments do not maintain context of the active script hash. This leads
to widget state loss. This change adds the active script hash to the
fragment implementation on initialization, and runs the same fragment.

## Testing Plan

- Added Unit tests to show the run_with_active_script_hash is called
correctly.
- Added an E2E test to show widget state is not lost.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady commented:
The hello_app test is failing the changed files test due to the hacky approach of setting up the hello files. The changed files does not retry, and so the test is a little flaky. This does succeed in the full playwright suite. We will address the hello app next, but going to merge here.

onewesong commented:
With this pull request getting merged, when can we expect this feature to become available for production use?

sfc-gh-jcarroll commented:
@onewesong we're planning a full release next week which will include it. You can use via `streamlit-nightly` now (although nightly is not fully certified for production use)
